src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                                        newArgumentExpression = newArgumentExpression.ReplaceNode(newArgumentExpression.GetCurrentNode(variable), parenthesizedGeneratedExpression);
+                                    }
+                                }
+                            }
+
+                            var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+                            var allArguments = AddArgumentToArgumentList(updatedInvocationArguments,
+                                newArgumentExpression.WithAdditionalAnnotations(Formatter.Annotation), insertionIndex, parameterName, parameterIsNamed);
+                            var newInvo = editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+                            return newInvo;
+                        });
+                    }
+
+                    if (document.Id == semanticDocument.Document.Id)
+                    {
+                        UpdateExpressionInOriginalFunction(semanticDocument, expressionCopy, oldMethodDeclaration, parameterName, editor, allOccurrences, cancellationToken);
+
+                        var parameterType = semanticDocument.SemanticModel.GetTypeInfo(expressionCopy, cancellationToken).Type ?? semanticDocument.SemanticModel.Compilation.ObjectType;
+                        var refKind = syntaxFacts.GetRefKindOfArgument(expressionCopy);
+                        var parameter = generator.ParameterDeclaration(name: parameterName, type: generator.TypeExpression(parameterType), refKind: refKind);
+                        editor.InsertParameter(oldMethodDeclaration, insertionIndex, parameter);
+                    }
+
+                    var newRoot = editor.GetChangedRoot();
+                    modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(firstCallSite.Id, newRoot);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter
+        /// </summary>
+        public static Dictionary<TIdentifierNameSyntax, IParameterSymbol> MapExpressionToParameters(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = document.SemanticModel;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol != null && symbol is IParameterSymbol parameterSymbol)
+                {
+                    if (!nameToParameterDict.ContainsKey(variable))
```

Comment:
how could this ever be true?

Reply:
Expression is: x + x + y
It adds the parameter corresponding to x, don't want it to add again once it hits the second x

Reply:
but the second 'x' will return 'false' for ContainsKey, so you will add it :)  (unless i'm horriblymisreading things).

Reply:
![image](https://user-images.githubusercontent.com/40616383/112236340-820fea80-8bfd-11eb-9438-84558c3b13a5.png)
I think you have it backwards

Reply:
> don't want it to add again once it hits the second x

but the two x's are are different objects.  so neither will result in ContainsKey for hte other.

Reply:
Sorry, I remember why I added it. It was for the params case in which the identifiers were actually the same object. Since I removed the params case, I can remove it